plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/transformers/BeanFactoryTransformer.java

```diff
@@ -19,34 +19,74 @@
-        clazz.addMethod(CtNewMethod.make("public void hotswapAgent$clearDestroyBean() { hotswapAgent$destroyBean.clear(); }", clazz));
+        clazz.addMethod(CtNewMethod.make(
+            "public boolean hotswapAgent$isDestroyedBean(String beanName) { return hotswapAgent$destroyBean.contains"
+                + "(beanName); }",
+            clazz));
+        clazz.addMethod(CtNewMethod.make(
+            "public void hotswapAgent$destroyBean(String beanName) { hotswapAgent$destroyBean.add(beanName); }",
+            clazz));
+        clazz.addMethod(
+            CtNewMethod.make("public void hotswapAgent$clearDestroyBean() { hotswapAgent$destroyBean.clear(); }",
+                clazz));
 
-        CtMethod destroySingletonMethod = clazz.getDeclaredMethod("destroySingleton", new CtClass[]{classPool.get(String.class.getName())});
-        destroySingletonMethod.insertAfter(BeanFactoryProcessor.class.getName() + ".postProcessDestroySingleton($0, $1);");
+        CtMethod destroySingletonMethod = clazz.getDeclaredMethod("destroySingleton",
+            new CtClass[] {classPool.get(String.class.getName())});
+        destroySingletonMethod.insertAfter(
+            BeanFactoryProcessor.class.getName() + ".postProcessDestroySingleton($0, $1);");
     }
 
     @OnClassLoadEvent(classNameRegexp = "org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory")
-    public static void registerAbstractAutowireCapableBeanFactory(ClassLoader appClassLoader, CtClass clazz, ClassPool classPool) throws NotFoundException, CannotCompileException {
+    public static void registerAbstractAutowireCapableBeanFactory(ClassLoader appClassLoader, CtClass clazz,
+        ClassPool classPool) throws NotFoundException, CannotCompileException {
+        clazz.addField(CtField.make("private static final org.hotswap.agent.logging.AgentLogger $$ha$LOGGER = " +
+                "org.hotswap.agent.logging.AgentLogger.getLogger(org.springframework.beans.factory.annotation"
+                + ".InitDestroyAnnotationBeanPostProcessor.class);",
+            clazz));
 
-        CtMethod createBeanMethod = clazz.getDeclaredMethod("createBean", new CtClass[]{
-                classPool.get(String.class.getName()),classPool.get("org.springframework.beans.factory.support.RootBeanDefinition"),
-                classPool.get(Object[].class.getName())});
+        CtMethod createBeanMethod = clazz.getDeclaredMethod("createBean", new CtClass[] {
+            classPool.get(String.class.getName()),
+            classPool.get("org.springframework.beans.factory.support.RootBeanDefinition"),
+            classPool.get(Object[].class.getName())});
         createBeanMethod.insertAfter(BeanFactoryProcessor.class.getName() + ".postProcessCreateBean($0, $1, $2);");
 
+        // try catch for custom init method
+        CtMethod invokeCustomInitMethod = clazz.getDeclaredMethod("invokeCustomInitMethod",
```

Comment:
should we check if 'invokeCustomInitMethod' and 'invokeInitMethod' are null before enhancing them, from the perspective of defensive programming?

Reply:
I have check the code from spring 3.2 to spring 6.x, both of these methods are exist.